edn/src/edn.rustpeg

```diff
@@ -252,6 +282,210 @@ pub entity -> Entity<ValueAndSpan>
     = __ "[" __ op:(op) __ e:(entity_place) __ a:(forward_entid)  __ v:(value_place) __  "]" __ { Entity::AddOrRetract { op, e: e, a: AttributePlace::Entid(a), v: v } }
     / __ "[" __ op:(op) __ e:(value_place)  __ a:(backward_entid) __ v:(entity_place) __ "]" __ { Entity::AddOrRetract { op, e: v, a: AttributePlace::Entid(a), v: e } }
     / __ map:map_notation __ { Entity::MapNotation(map) }
+    / #expected("entity")
 
 pub entities -> Vec<Entity<ValueAndSpan>>
     = __ "[" __ es:(entity*) __ "]" __ { es }
+
+// Query parser starts here.
+//
+// We expect every rule except the `raw_*` rules to eat whitespace
+// (with `__`) at its start and finish.  That means that every string
+// pattern (say "[") should be bracketed on either side with either a
+// whitespace-eating rule or an explicit whitespace eating `__`.
+
+query_function -> query::QueryFunction
```

Comment:
Do we have a documented grammar definition somewhere for reference? It would be much easier to skim through for reference than this syntax (which I realize isn't intended for direct human consumption).

Reply:
We run off of https://docs.datomic.com/on-prem/query.html#grammar -- except where it's subtly wrong.  And we do have some extensions.